File: src/main/Java/boj_step6/boj1157.java
import java.io.*;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class boj1157 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        char[] chars = reader.readLine().toUpperCase().toCharArray();
        HashMap map = new HashMap<Character, Integer>();
        for (char c: chars) {
            if (map.containsKey(c)) map.put(c, (int)map.get(c) + 1);
            else map.put(c, 1);
        }
        int maxValue = (int) Collections.max(map.values());
        Optional<Map.Entry<Character, Integer>> maxEntry = map.entrySet().stream().max(Map.Entry.comparingByValue());
        long count = map.values().stream().filter( value -> (int)value == maxValue ).count();
        if (count > 1) writer.append("?");
        else writer.append(String.valueOf(maxEntry.get().getKey()));
        writer.close();
    }
}

File: src/main/kotlin/boj1157.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import kotlin.math.max

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val s = r.readLine()!!
    val map = mutableMapOf<Char, Int>()
    s.forEach {
        val char = it.toLowerCase()

        if (map.containsKey(char)) map[char] = map[char]!!.plus(1)
        else map.set(char, 1)
    }
    val max = map.maxBy { it.value }

    if(map.filter { it.value == max!!.value }.count() > 1)
        w.write("?")
    else w.write(max!!.key.toUpperCase().toString())
    w.close()
}